chain/epoch_manager/src/lib.rs

```diff
@@ -410,6 +411,12 @@ impl EpochManager {
             )
         };
         let next_next_epoch_config = self.config.for_protocol_version(next_version);
+        #[cfg(feature = "sandbox")]
+        {
+            *next_epoch_info.epoch_height_mut() =
+                block_info.height() / next_next_epoch_config.epoch_length;
+        }
```

Comment:
Could you describe why this change is necessary? In general the relationship described here does not always hold.

Reply:
this updates the epoch height to after fast-forwarding the block height. This was the only way I could feasibly update it since `epoch_manager` is separate from everything else. Also, what cases does this not hold for? This is just for sandbox so it might not be necessary to account for them

Reply:
> Also, what cases does this not hold for?

While it is guaranteed that the block height of the end of an epoch is at least `epoch_length` + the height of the block in the beginning of the epoch, there is no guarantee on the upper bound. In fact, it is possible for a block height in one epoch to be arbitrarily large. Could you let me know the reason for modifying epoch height? What guarantees do you need to have here?

Reply:
> there is no guarantee on the upper bound. In fact, it is possible for a block height in one epoch to be arbitrarily large

This should be fine since when we receive a fast-forward request, it happens in the middle of an epoch and the newly forward block height would trigger updating the epoch-height at an arbitrary block height.

The guarantees needed for epoch height updates would be what you mentioned at the end of an epoch:
- block-height lower bound: `epoch_length + height of block in beginning of the epoch`
- block-height upper bound: unbounded

> Could you let me know the reason for modifying epoch height?

This is needed because what I mentioned earlier in the other post about producing blocks and simulating those. So after producing X amount of blocks (fast-forwarding), the user of a contract would intuitively assume that the epoch height also got updated to reflect X blocks being produced.

Also, correct me on any of these things, but that's my thinking at least

Reply:
> So after producing X amount of blocks (fast-forwarding), the user of a contract would intuitively assume that the epoch height also got updated to reflect X blocks being produced.

Actually this should be done automatically as long as block height is more than `epoch_length` + height of block in the beginning of the epoch. I don't think we need custom logic here.

Reply:
hmm interesting. But wouldn't that mean the epoch-height is only incremented by 1? There was only logic I saw that incremented the epoch-height by 1 when the next epoch was in view: https://github.com/near/nearcore/blob/31493321477fac77b466305c019117f61c930b63/chain/epoch_manager/src/validator_selection.rs#L194

Reply:
Yes. Do you require it to be incremented by more than 1? If so, what is the use case?

Reply:
not so much that it needs to be incremented more than 1 right there, but the custom logic is what I'm saying allows for us to skip to the designated epoch height after fast forwarding.

> Actually this should be done automatically as long as block height is more than epoch_length + height of block in the beginning of the epoch. I don't think we need custom logic here.

But now I'm not sure what you're referring to then with the above statement. Is there some other logic that would allow us to jump to the new epoch height after fast forwarding? I didn't see anything and epoch height remained the same as if fast-forwarding the block height never happened

Reply:
I think you just need to produce three more blocks to end the epoch, which could be implemented as part of the sandbox

Reply:
It shouldn't need to produce more blocks to end the epoch though, but regardless I tested it without the custom logic and it's not producing what's expected after fast forwarding. I'm getting epoch height of 3, when it should be 503, after waiting for roughly 50-60 blocks to be produced. You can see for yourself with the following diff and running `python3 pytests/tests/sandbox/fast_forward.py`:
```diff
diff --git a/chain/epoch_manager/src/lib.rs b/chain/epoch_manager/src/lib.rs
index b54af9e47..b1406c930 100644
--- a/chain/epoch_manager/src/lib.rs
+++ b/chain/epoch_manager/src/lib.rs
@@ -411,18 +411,6 @@ impl EpochManager {
             )
         };
         let next_next_epoch_config = self.config.for_protocol_version(next_version);
-        #[cfg(feature = "sandbox")]
-        {
-            // explaination on calculation:
-            // (block_height / length) gives us epoch_height, but we index from 1, so +1 at the end.
-            // Note, this is a retroactive update for sandbox since we can receive fast-forward
-            // requests in the middle of an epoch. block_height is different based on when it gets
-            // called. When fast-forwarded, we need to retroactively adjust the epoch height, but
-            // with a normal `finalize_epoch`, it's the start of the epoch's next block height, so
-            // (height - 1) is needed after adding +1 at the end.
-            *next_epoch_info.epoch_height_mut() =
-                ((block_info.height() - 1) / next_next_epoch_config.epoch_length) + 1;
-        }
 
         let next_next_epoch_info = match proposals_to_epoch_info(
             next_next_epoch_config,
@@ -460,12 +448,6 @@ impl EpochManager {
         // where epoch_id of it is the hash of last block in this epoch (T).
         self.save_epoch_info(store_update, &next_next_epoch_id, next_next_epoch_info)?;
 
-        // Store next_epoch info again since sandbox changes it. Need to retroactively update
-        // since a fast-forward request can happen in the middle of an epoch.
-        if cfg!(feature = "sandbox") {
-            self.save_epoch_info(store_update, &next_epoch_id, next_epoch_info)?;
-        }
-
         Ok(())
     }
 
diff --git a/pytest/tests/sandbox/fast_forward.py b/pytest/tests/sandbox/fast_forward.py
index 569981ab3..87287fd40 100644
--- a/pytest/tests/sandbox/fast_forward.py
+++ b/pytest/tests/sandbox/fast_forward.py
@@ -61,4 +61,9 @@ assert min_forwarded_time < latest < max_forwarded_time
 
 # Check to see that the epoch height has been updated correctly:
 epoch_height = nodes[0].get_validators()['result']['epoch_height']
-assert epoch_height > BLOCKS_TO_FASTFORWARD / EPOCH_LENGTH
+# assert epoch_height > BLOCKS_TO_FASTFORWARD / EPOCH_LENGTH
+print(f"-> epoch_height {epoch_height}")
+
+utils.wait_for_blocks(nodes[0], target=BLOCKS_TO_FASTFORWARD + 50)
+epoch_height = nodes[0].get_validators()['result']['epoch_height']
+print(f"-> epoch_height {epoch_height}")
```
And after including the custom logic again, the python tests show the correct epoch height of 503, so it's definitely needed here